local-cluster/tests/local_cluster.rs

```diff
@@ -1680,7 +1601,282 @@ fn test_validator_saves_tower() {
+//    so `A` will eventually pick the fork `C` is on.
+//
+//    Furthermore `B`'s vote on `S3` is not observable because there are no
+//    descendants of slot `S3`, so that fork will not be chosen over `C`'s fork
+//
+// With the persisted tower:
+//    `A` should not be able to generate a switching proof.
+//
+fn do_test_optimistic_confirmation_violation_with_or_without_tower(with_tower: bool) {
+    solana_logger::setup();
+
+    // First set up the cluster with 4 nodes
+    let slots_per_epoch = 2048;
+    let node_stakes = vec![31, 36, 33, 0];
+
+    // Each pubkeys are prefixed with A, B, C and D.
+    // D is needed to avoid NoPropagatedConfirmation erorrs
+    let validator_keys = vec![
+        "28bN3xyvrP4E8LwEgtLjhnkb7cY4amQb6DrYAbAYjgRV4GAGgkVM2K7wnxnAS7WDneuavza7x21MiafLu1HkwQt4",
+        "2saHBBoTkLMmttmPQP8KfBkcCw45S5cwtV3wTdGCscRC8uxdgvHxpHiWXKx4LvJjNJtnNcbSv5NdheokFFqnNDt8",
+        "4mx9yoFBeYasDKBGDWCTWGJdWuJCKbgqmuP8bN9umybCh5Jzngw7KQxe99Rf5uzfyzgba1i65rJW4Wqk7Ab5S8ye",
+        "3zsEPEDsjfEay7te9XqNjRTCE7vwuT6u4DHzBJC19yp7GS8BuNRMRjnpVrKCBzb3d44kxc4KPGSHkCmk6tEfswCg",
+    ]
+    .iter()
+    .map(|s| (Arc::new(Keypair::from_base58_string(s)), true))
+    .take(node_stakes.len())
+    .collect::<Vec<_>>();
+    let validators = validator_keys
+        .iter()
+        .map(|(kp, _)| kp.pubkey())
+        .collect::<Vec<_>>();
+    let (validator_a_pubkey, validator_b_pubkey, validator_c_pubkey) =
+        (validators[0], validators[1], validators[2]);
+
+    let config = ClusterConfig {
+        cluster_lamports: 100_000,
+        node_stakes: node_stakes.clone(),
+        validator_configs: vec![ValidatorConfig::default(); node_stakes.len()],
+        validator_keys: Some(validator_keys),
+        slots_per_epoch,
+        stakers_slot_offset: slots_per_epoch,
+        skip_warmup_slots: true,
+        ..ClusterConfig::default()
+    };
+    let mut cluster = LocalCluster::new(&config);
+
+    let base_slot = 26; // S2
+    let next_slot_on_a = 27; // S3
+    let truncated_slots = 100; // just enough to purge all following slots after the S2 and S3
+
+    // Immediately kill validator C
+    let validator_c_info = cluster.exit_node(&validator_c_pubkey);
+
+    // Step 1:
+    // Let validator A, B, (D) run for a while.
+    let client = cluster.get_validator_client(&validator_a_pubkey).unwrap();
+    loop {
+        sleep(Duration::from_millis(100));
+        let highest_bank = client
+            .get_slot_with_commitment(CommitmentConfig::recent())
+            .unwrap();
+        if highest_bank >= next_slot_on_a {
+            break;
+        }
+    }
+    let validator_a_info = cluster.exit_node(&validator_a_pubkey);
+    let client = cluster.get_validator_client(&validator_b_pubkey).unwrap();
+    loop {
+        sleep(Duration::from_millis(100));
+        let highest_bank = client
+            .get_slot_with_commitment(CommitmentConfig::recent())
```

Comment:
Is there a way to assert here that validator `b` actually voted on `next_slot_on_a` (maybe a variation of `last_vote_in_tower()` that checks `tower_contains_vote()`? 

I think CommitmentConfig::recent() implies the validator voted

Reply:
yeah, this suggestion makes sense. well, I was just lazy... ;)